### PR TITLE
Changes to Accommodate new Employee Mappings Table

### PR DESCRIPTION
**Pull request by ashwin1111**

### CODE REVIEW

File: sql/scripts/011-delete-emp-map-settings.sql [hunk 1]
```diff
@@ -0,0 +1,6 @@
+-- Script to delete employee mapping setting data since they're moved to workspace general settings
```
Comment (ashwin1111) [inline sql/scripts/011-delete-emp-map-settings.sql:1]: TODO: script to migrate employee mapping data to new table

File: apps/quickbooks_online/tasks.py [hunk 2]
```diff
@@ -114,6 +114,7 @@ def create_or_update_employee_mapping(expense_group: ExpenseGroup, qbo_connectio
                     create=True
                 )
 
+            # TODO: fix this, create a helper function in Accounting Mappings
```
Comment (ashwin1111) [inline apps/quickbooks_online/tasks.py:117]: TODO

File: apps/mappings/tasks.py [hunk 3]
```diff
@@ -242,6 +243,132 @@ def schedule_categories_creation(import_categories, workspace_id):
             schedule.delete()
 
 
+def get_existing_source_and_mappings(destination_type: str, workspace_id: int):
```
Comment (ashwin1111) [inline apps/mappings/tasks.py:246]: Auto map changes based on new table

File: apps/mappings/tasks.py [hunk 4]
```diff
@@ -242,6 +243,132 @@ def schedule_categories_creation(import_categories, workspace_id):
...
+                source_value = source_attribute.value
+            elif employee_mapping_preference == 'NAME':
+                source_value = source_attribute.detail['full_name']
+            elif employee_mapping_preference == 'EMPLOYEE_CODE':
+                source_value = source_attribute.detail['employee_code']
+
+            # Checking exact match
+            if source_value.lower() in destination_id_value_map:
+                destination = {}
+                if destination_type == 'EMPLOYEE':
```
Comment (ashwin1111) [inline apps/mappings/tasks.py:285]: Auto map changes based on new table

File: apps/mappings/tasks.py [hunk 5]
```diff
@@ -242,6 +243,132 @@ def schedule_categories_creation(import_categories, workspace_id):
...
+                    destination['destination_employee_id'] = destination_id_value_map[source_value.lower()]
+                elif destination_type == 'VENDOR':
+                    destination['destination_vendor_id'] = destination_id_value_map[source_value.lower()]
+                elif destination_type == 'CREDIT_CARD_ACCOUNT':
+                    destination['destination_card_account_id'] = destination_id_value_map[source_value.lower()]
+
+                mapping_batch.append(
+                    EmployeeMapping(
```
Comment (ashwin1111) [inline apps/mappings/tasks.py:293]: Auto map changes based on new table

File: apps/mappings/tasks.py [hunk 6]
```diff
@@ -242,6 +243,132 @@ def schedule_categories_creation(import_categories, workspace_id):
...
+                        source_employee_id=source_attribute.id,
+                        workspace_id=workspace_id,
+                        **destination
+                    )
+                )
+
+    return mapping_batch
+
+
+def create_mappings_and_update_flag(mapping_batch: list, set_auto_mapped_flag: bool = True):
```
Comment (ashwin1111) [inline apps/mappings/tasks.py:303]: Almost same as before

File: apps/mappings/tasks.py [hunk 7]
```diff
@@ -242,6 +243,132 @@ def schedule_categories_creation(import_categories, workspace_id):
...
+            )
+
+        if expense_attributes_to_be_updated:
+            ExpenseAttribute.objects.bulk_update(
+                expense_attributes_to_be_updated, fields=['auto_mapped'], batch_size=50)
+
+    return mappings
+
+
+def auto_map_employees(destination_type: str, employee_mapping_preference: str, workspace_id: int):
```
Comment (ashwin1111) [inline apps/mappings/tasks.py:324]: Same as before

File: apps/mappings/tasks.py [hunk 8]
```diff
@@ -285,6 +412,55 @@ def schedule_auto_map_employees(employee_mapping_preference: str, workspace_id:
...
+    employee_source_attributes = ExpenseAttribute.objects.filter(
+        attribute_type='EMPLOYEE', workspace_id=workspace_id
+    ).all()
+
+    default_destination_attribute = DestinationAttribute.objects.filter(
+        destination_id=default_ccc_account_id, workspace_id=workspace_id, attribute_type='CREDIT_CARD_ACCOUNT'
+    ).first()
+
+    existing_source_ids, existing_mappings = get_existing_source_and_mappings('CREDIT_CARD_ACCOUNT', workspace_id)
+    existing_mappings_map = {mapping.source_employee_id: mapping.id for mapping in existing_mappings}
```
Comment (ashwin1111) [inline apps/mappings/tasks.py:430]: Auto map changes based on new table

File: apps/mappings/tasks.py [hunk 9]
```diff
@@ -285,6 +412,55 @@ def schedule_auto_map_employees(employee_mapping_preference: str, workspace_id:
...
+
+    mapping_creation_batch = []
+    mapping_updation_batch = []
```
Comment (ashwin1111) [inline apps/mappings/tasks.py:433]: Auto map changes based on new table

File: apps/mappings/tasks.py [hunk 10]
```diff
@@ -242,6 +243,132 @@ def schedule_categories_creation(import_categories, workspace_id):
...
+            value_to_be_appended = destination_employee.detail['email'].replace('*', '')
+        elif employee_mapping_preference in ['NAME', 'EMPLOYEE_CODE']:
+            value_to_be_appended = destination_employee.value.replace('*', '')
+
+        if value_to_be_appended:
+            attribute_values = '{}|{}'.format(attribute_values, value_to_be_appended.lower())
+            destination_id_value_map[value_to_be_appended.lower()] = destination_employee.id
+
+    if employee_mapping_preference == 'EMAIL':
+        filter_on = 'value__iregex'
```
Comment (ashwin1111) [inline apps/mappings/tasks.py:350]: TODO: remove all regex stuffs + add pagination

**Review by Shwetabhk — APPROVED** [review]

**Pull request merged**